src/utils/colorManipulation.test.ts

```diff
@@ -0,0 +1,231 @@
+
+      // Determine what my test should return by hand
+      const expected = [255, 0, 0, 255];
+
+      // Run my code and get it's actual output
+      const actual = getAverageColor(frame, 0, 0, canvas.height, canvas.width);
+
+      // Use an expect to actual run assertions with the matchers that jest provides
+      expect(actual).toEqual(expected);
+    });
+
+    it("averages two colors", () => {
+      ctx.fillStyle = "rgb(0, 0, 0)";
+      ctx.fillRect(0, 0, canvas.width / 2, canvas.height);
+      ctx.fillStyle = "rgb(255, 255, 255)";
+      ctx.fillRect(canvas.width / 2, 0, canvas.width, canvas.height);
+
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      const expected = [128, 128, 128, 255];
+      const actual = getAverageColor(frame, 0, 0, canvas.height, canvas.width);
+      expect(actual).toEqual(expected);
+    });
+
+    it("averages four colors", () => {
+      // top left quadrant
+      ctx.fillStyle = "rgb(255, 0, 0)";
+      ctx.fillRect(0, 0, canvas.width / 4, canvas.height / 4);
+
+      // top right quadrant
+      ctx.fillStyle = "rgb(0, 255, 0)";
+      ctx.fillRect(canvas.width / 2, 0, canvas.width / 4, canvas.height / 4);
+
+      // bottom left quadrant
+      ctx.fillStyle = "rgb(0, 0, 255)";
+      ctx.fillRect(0, canvas.height / 2, canvas.width / 4, canvas.height / 4);
+
+      // bottom right quadrant
+      ctx.fillStyle = "rgb(0, 0, 0)";
+      ctx.fillRect(
+        canvas.width / 2,
+        canvas.height / 2,
+        canvas.width / 4,
+        canvas.height / 4,
+      );
+
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      const expected = [16, 16, 16, 255];
+      const actual = getAverageColor(frame, 0, 0, canvas.height, canvas.width);
+      expect(actual).toEqual(expected);
+    });
+
+    it("does not average colors outside of the region", () => {
+      // draw the bg red
+      ctx.fillStyle = "rgb(255, 0, 0)";
+      ctx.fillRect(0, 0, canvas.width, canvas.height);
+
+      // draw a green square in the center
+      ctx.fillStyle = "rgb(0, 255, 0)";
+      ctx.fillRect(
+        canvas.width / 4,
+        canvas.width / 4,
+        canvas.width / 2,
+        canvas.height / 2,
+      );
+
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      const expected = [0, 255, 0, 255];
+      const actual = getAverageColor(
+        frame,
+        canvas.height / 4,
+        canvas.width / 4,
+        (canvas.height * 3) / 4,
+        (canvas.width * 3) / 4,
+      );
+      expect(actual).toEqual(expected);
+    });
+  });
+
+  describe("computeDivisions", () => {
+    it("return even divisions", () => {
+      const verticalDivisions = 4,
+        horizontalDivisions = 2;
+      const videoDimensions: Dimensions = { height: 100, width: 100 };
+
+      // need color to complete the type, but not being tested here
+      const expected: Array<Division> = [
```

Comment:
this portion took me a bit to comprehend, haha, heres how i drew it out as: (https://github.com/user-attachments/assets/b1bc0e44-68da-477a-b2f9-479902521e3c)

Reply:
thanks for feedback i think i can refactor this to be more readable with a loop. your drawing is spot on tho!